src/Uno.UI.Composition/Composition/Visual.skia.cs

```diff
@@ -177,23 +184,60 @@ internal virtual void Render(in DrawingSession parentSession)
 			return;
 		}
 
-		using var session = BeginDrawing(in parentSession);
-		Draw(in session);
+		using var session = parentSession.WithOpacity(Opacity);
+		SetupLocalCoordinatesAndClip(in session);
```

Comment:
Why did you remove the `BeginDrawing`? This correspond to the logic of that method and the `PaintingSession` was designed to abstract the notion/count of `Canvas.Save` and `Canvas.Restore`.

Here you create a new session in the `WIthOpacity` and then do the `Save` in `SetupLocalCoordinatesAndClip`. Invoking one of the 2 method without invoking the other one will cause issue. Please merge them back. 

(And it was the purpose of the `PushOpacity`: setuping current session without creating a new one ... but its probably no longer relevant since I think we does that only here).

Reply:
I definitely agree with this. In fact, I just went a step further (I most likely went way too far with this actually): a creating and disposing a `PaintingSession` should not concern anything outside the `Visual` that creates it to start a new "local session". I moved `PaintingSession` to be nested inside `Visual` and tried to make it work in an RAII-ish way. Let me know if you think it's too over the top.